hal/src/thumbv7em/nvm.rs

```diff
@@ -0,0 +1,515 @@
+//! # Non-volatile Memory Controller
+//!
+//! This module allows users to interact with non-volatile memory controller.
+//!
+//! NVMCTRL is an intermediary between memory buses and physical non-volatile
+//! memory. It provides means of managing a flash memory content, its properties
+//! (cache, wait states, bootloader blocks protection), power management and
+//! address remapping if necessary (in case bank mechanism is used). It also
+//! provides an indirection mechanism to achieve non-volatile RAM-like memory
+//! within last sectors of a physical flash (More in [`smart_eeprom`] module).
+//!
+//! NVM supports splitting flash into two sections (opt-in feature) called
+//! banks. Bank considered active is mapped to _virtual_ address `0x0`, meaning
+//! it contains currently executed application. Through NVM command & control
+//! interface, banks can be swapped and MCU reset, so the firmware from the
+//! other bank will run after restart.
+//!
+//! Module features:
+//! - Erase & write over non-volatile memory in a device.
+//! - Swap banks
+#![warn(missing_docs)]
+
+pub mod smart_eeprom;
+
+pub use crate::target_device::nvmctrl::ctrla::PRM_A;
+use crate::target_device::nvmctrl::ctrlb::CMD_AW;
+use crate::target_device::NVMCTRL;
+use core::ops::Range;
+
+use bitfield::bitfield;
+
+#[inline(always)]
+fn retrieve_flash_size() -> u32 {
+    static mut FLASHSIZE: Option<u32> = None;
```

Comment:
I think this option of yours is better than regular atomics (if I understood @TDHolmes correctly). `Option<AtomicU32> does not guarantee that `Option` access is atomic, I guess?

Reply:
Correct, I was thinking move away from `Option<T>` completely in favor of `AtomicU32` with it initialized to zero, and the FW knows that's an invalid value. This removes the need for it to be `mut` and guarantees atomicity instead of implicit atomicity of `Option<NonZeroU32>` which only happens to be atomic due to it being represented by a `u32` under the hood and how ARM works

Reply:
I don't feel messing with ordering in `AtomicU32`, I don't see can it improve "atomic" properties of this function, as you have to read from it first, have a branching (non-zero -> return; zero -> read from HW registers and calculate value), and write to it. I don't have any expertise regarding  memory read/write ordering so I pass here. If some wants to fix it at any point in time, feel welcome.